### PR TITLE
Implement svn+ssh for Gitea

### PR DESCRIPTION
**Pull request by zeripath**
Hi!

Here is an implementation for making git-as-svn be able to work with ssh on gitea. This is work in progress.

### CODE REVIEW

**Comment by zeripath** [comment]
OK, so this has not been completely tested as yet, but I thought I'd show you what I've been working on. Yes some parts are hideous but... it's the best that can be done at the present I think for gitea.

Issues:
1. The bash tunnel needs to be checked to ensure that the file-descriptors close properly, and that if the client or server disconnects abruptly these aren't stuck around. I think this also needs to be checked so that the client cannot DDOS the server. I wrote the tunnel in bash because the gitea docker does not have python installed, but I guess it does not have java installed either.
2. Gitea <v1.7 will not have a reasonable keys api - hence the horrible SSHKeysWatcher implementation - I have a pull request for a better API in gitea but it's not expected to be approved prior to the release of 1.6.0.
3. It's somewhat horrible having to parse and rewrite the authorized_keys file but gitea has a standardised way of writing to this file, and thus the rewrite can be very simple. @fcharlie appeared to have a technique for managing the internal ssh server, but when using an external ssh server, without having to then manage SVN SSH keys separately this seems like the only technique.

**Comment by slonopotamus** [comment]
Overall thoughts:

1. I didn't read bash thingie carefully, it's 1am here ;)
2. Keys thread is... ugh. Need gitea keys api to fix it.
3. Will also need to package scripts somewhere when we make a release
4. Actually, since you're the only gitea + git-as-svn user on the planet currently, I don't see any strong reasons against taking this if it works for you.
5. I hope you understand that gitea stuff neither has tests nor easy to setup manually, so we can unintentionally damage it a bit when changing core git-as-svn code.

What additional changes do you want to do before decide this is ready to be merged?

File: src/main/java/svnserver/ext/gitea/keys/SSHDirectoryWatcher.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      }
+    });
+
+    // OK now we know about which keys are there.
+    // So we tell our keys mapper...
+    this.mapper.setKeys(keysSet);
+  }
+
+  public void close() {
+    this.interrupt();
```
[the remark this answers is not in the document]
Comment (slonopotamus) [reply src/main/java/svnserver/ext/gitea/keys/SSHDirectoryWatcher.java]: Okay, actually what I said is also isn't 100% racy-free, `interrupt` could happen between `!stopped` check and blocking on watchService.

**Comment by zeripath** [comment]
Heya Slonoptamus! 

Your comment about tests and ease to set-up manually made me think about how to create integration tests so I've been busy adding some integration tests.

It's strange what you said about setting up gitea manually though, as I thought gitea was a lot easier to set up than gitlab! Are you struggling with adding java to the gitea docker?

```
apk add openjdk8-jre-base
```

will add java so you can run git-as-svn from within the docker.

File: src/main/java/svnserver/ext/gitea/keys/SSHDirectoryWatcher.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeripath) [reply src/main/java/svnserver/ext/gitea/keys/SSHDirectoryWatcher.java]: Would it not be sufficient to just check if the thread isInterrupted() at a few opportune points? Like at the start of the while and perhaps just after watchService.take()

File: src/main/java/svnserver/ext/gitea/keys/KeysMapper.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  public String getUserName(@NotNull String keyId) {
+    return keyMap.get(keyId);
+  }
+
+  @NotNull
+  public String getSecretToken() {
+    return secretToken;
+  }
+
+  public void run() {
```
[the remark this answers is not in the document]
Comment (zeripath) [reply src/main/java/svnserver/ext/gitea/keys/KeysMapper.java]: The second thread decouples the collection of the keys from reprocessing the authorized_keys file - meaning that if there are rapid changes to the authorized_keys file you only collect them every 15 seconds.

File: src/main/java/svnserver/ext/gitea/keys/SSHDirectoryWatcher.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import svnserver.context.Shared;
+
+/**
+ * SSHDirectoryWatcher.
+ *
+ * @author Andrew Thornton <zeripath@users.noreply.github.com>
+ */
+public class SSHDirectoryWatcher extends Thread implements Shared {
```
[the remark this answers is not in the document]
Comment (zeripath) [reply src/main/java/svnserver/ext/gitea/keys/SSHDirectoryWatcher.java]: Yeah. It's just a tech preview to show that it is possible.

File: src/main/java/svnserver/ext/gitea/keys/KeysMapper.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+          }
+        }
+      }
+    } catch (InterruptedException e) {
+      return;
+    }
+  }
+
+  public void close() {
+    this.interrupt();
```
[the remark this answers is not in the document]
Comment (zeripath) [reply src/main/java/svnserver/ext/gitea/keys/KeysMapper.java]: See above. I've changed the implementation slightly to check isInterrupted meaning that I don't think a volatile is needed.

**Comment by zeripath** [comment]
I've made major changes to the bash script that stands in the middle of the subversion stream.  This should handle errors a lot better and I think it's a bit clearer. 

I'm not sure how to create an integration test for this. Any ideas? 

What do you think of the tunnel approach? It's a little hacky, but I think the JVM startup times likely too long to be appropriate and without adding C or python to the codebase it seems that bash is the best on offer. The tunnel approach would be extensible to Gitlab, however, Gitlab wouldn't require the SSHDirectoryWatcher approach as Gitlab actually uses a AuthorizedKeysCommand which could be leveraged.

**Comment by slonopotamus** [comment]
Bash is fine. WRT tests - we have a fundamental problem that when we run tests, git-as-svn is _outside_ of docker. So, for example, we cannot properly test commiting to repo:
a) we actually don't have repo, it is inside container (okay, this is fixable)
b) at least for GitLab we cannot properly call postcommit hook because it won't find Gitlab itself

There are two possible solutions:
a) Learn to operate in completely remote mode (it means we need to pull/push to upstream git repo and proxy all LFS stuff)
b) Make some kind of heavy-weight integration tests where we install git-as-svn into docker, setup & start it there)

Both options are definitely out of scope of this PR.

**Comment by slonopotamus** [comment]
So. Do you want me to merge this?

**Comment by zeripath** [comment]
Hmm, not yet.

I think it probably still needs some more documentation, and some packaging.

The other thing is whether it would be worth implementing svn+ssh for
gitlab as part of the PR, so it didn't feel like it was a total outlier.


-- 
Andrew Thornton
art27@cantab.net

**Comment by slonopotamus** [comment]
I merged other two of your PRs, this one conflicts with `master` again a bit.

**Comment by zeripath** [comment]
Ah the joys of merging... I'll take a look at it tomorrow. Did you manage to get this working for you on gitlab? 

There's a possibility that go-gitea/gitea#4690 might be merged soon which would mean that things could be cleaned up a bit for gitea.

**Comment by zeripath** [comment]
Agh! I forgot to fix my changes. Just did it. I also put code in to manage Gitea v1.7.0+.

**Comment by slonopotamus** [comment]
@zeripath since go-gitea/gitea#4960 was merged long time ago, maybe we now can get rid of messing with `authorized_keys` file and work through Gitea API?

**Pull request merged**